src/saturn_engine/worker/services/usage_metrics.py

```diff
@@ -0,0 +1,158 @@
+SubmittingMessage = t.NewType("SubmittingMessage", ExecutableMessage)
+ExecutingMessage = t.NewType("ExecutingMessage", ExecutableMessage)
+
+T = t.TypeVar("T", bound=ExecutableMessage)
+U = t.TypeVar("U", bound=ExecutableMessage)
+
+
+class PipelineUsage(t.NamedTuple):
+    name: str
+    usage: float
+
+
+@dataclasses.dataclass
+class PipelineState:
+    messages: WeakSet[ExecutableMessage] = dataclasses.field(default_factory=WeakSet)
+    # The default_factory needs to use lambda otherwise it won't get overriden
+    # in tests by freezegun.
+    last_flushed_at: Nanoseconds = dataclasses.field(
+        default_factory=lambda: time.perf_counter_ns()
+    )
+    last_collected_at: Nanoseconds = dataclasses.field(
+        default_factory=lambda: time.perf_counter_ns()
+    )
+    usage: Nanoseconds = 0
+
+    def flush(self, *, now: Nanoseconds) -> None:
+        now = time.perf_counter_ns()
+        since_flush = now - self.last_flushed_at
+        self.usage += since_flush * len(self.messages)
+        self.last_flushed_at = now
+
+    def add(self, xmsg: ExecutableMessage) -> None:
+        self.flush(now=time.perf_counter_ns())
+        self.messages.add(xmsg)
+
+    def pop(self, xmsg: ExecutableMessage) -> None:
+        if xmsg in self.messages:
+            self.flush(now=time.perf_counter_ns())
+            self.messages.discard(xmsg)
+
+    def collect(self, *, now: Nanoseconds) -> float:
+        self.flush(now=now)
+        usage = self.usage / (now - self.last_collected_at)
+        self.last_collected_at = now
+        self.usage = 0
+        return usage
+
+
+@dataclasses.dataclass
+class StageState(t.Generic[T, U]):
+    pipelines: dict[PipelineName, PipelineState] = dataclasses.field(
+        default_factory=dict
+    )
+
+    def push(self, xmsg: T) -> None:
+        pipeline_state = self.pipelines.setdefault(
+            xmsg.message.info.name, PipelineState()
+        )
+        pipeline_state.add(xmsg)
+
+    def pop(self, xmsg: ExecutableMessage) -> U:
+        if pipeline_state := self.pipelines.get(xmsg.message.info.name):
+            pipeline_state.pop(xmsg)
+
+        return t.cast(U, xmsg)
+
+    def collect(self, *, now: Nanoseconds) -> t.Iterator[PipelineUsage]:
+        for pipeline, state in self.pipelines.items():
+            yield PipelineUsage(name=pipeline, usage=state.collect(now=now))
+
+
+@dataclasses.dataclass
+class StagesState:
+    polling: StageState[ExecutableMessage, PollingMessage] = dataclasses.field(
+        default_factory=StageState
+    )
+    scheduling: StageState[PollingMessage, SchedulingMessage] = dataclasses.field(
+        default_factory=StageState
+    )
+    submitting: StageState[SchedulingMessage, SubmittingMessage] = dataclasses.field(
+        default_factory=StageState
+    )
+    executing: StageState[SubmittingMessage, ExecutingMessage] = dataclasses.field(
+        default_factory=StageState
+    )
+
+
+class UsageMetrics(MinimalService):
```

Comment:
Do you have a quick example of how we use this in our own deployment?

Reply:
Just add it as a service in the config and everything work as-is.